arch/risc-v/src/bl602/bl602_i2c.c

```diff
@@ -0,0 +1,1049 @@
+    }
+}
+
+/****************************************************************************
+ * Name: bl602_i2c_setprd
+ *
+ * Description:
+ *   set i2c prd
+ *
+ ****************************************************************************/
+
+static void bl602_i2c_setprd(uint8_t phase)
+{
+  modifyreg32(BL602_I2C_PRD_START, I2C_PRD_START_CR_PRD_S_PH_0_MASK, phase);
+  modifyreg32(BL602_I2C_PRD_START,
+              I2C_PRD_START_CR_PRD_S_PH_1_MASK,
+              phase << I2C_PRD_START_CR_PRD_S_PH_1_SHIFT);
+  modifyreg32(BL602_I2C_PRD_START,
+              I2C_PRD_START_CR_PRD_S_PH_2_MASK,
+              phase << I2C_PRD_START_CR_PRD_S_PH_2_SHIFT);
+  modifyreg32(BL602_I2C_PRD_START,
+              I2C_PRD_START_CR_PRD_S_PH_3_MASK,
+              phase << I2C_PRD_START_CR_PRD_S_PH_3_SHIFT);
+
+  modifyreg32(BL602_I2C_PRD_STOP, I2C_PRD_STOP_CR_PRD_P_PH_0_MASK, phase);
+  modifyreg32(BL602_I2C_PRD_STOP,
+              I2C_PRD_STOP_CR_PRD_P_PH_1_MASK,
+              phase << I2C_PRD_STOP_CR_PRD_P_PH_1_SHIFT);
+  modifyreg32(BL602_I2C_PRD_STOP,
+              I2C_PRD_STOP_CR_PRD_P_PH_2_MASK,
+              phase << I2C_PRD_STOP_CR_PRD_P_PH_2_SHIFT);
+  modifyreg32(BL602_I2C_PRD_STOP,
+              I2C_PRD_STOP_CR_PRD_P_PH_3_MASK,
+              phase << I2C_PRD_STOP_CR_PRD_P_PH_3_SHIFT);
+
+  modifyreg32(BL602_I2C_PRD_DATA, I2C_PRD_DATA_CR_PRD_D_PH_0_MASK, phase);
+  modifyreg32(BL602_I2C_PRD_DATA,
+              I2C_PRD_DATA_CR_PRD_D_PH_1_MASK,
+              phase << I2C_PRD_DATA_CR_PRD_D_PH_1_SHIFT);
+  modifyreg32(BL602_I2C_PRD_DATA,
+              I2C_PRD_DATA_CR_PRD_D_PH_2_MASK,
+              phase << I2C_PRD_DATA_CR_PRD_D_PH_2_SHIFT);
+  modifyreg32(BL602_I2C_PRD_DATA,
+              I2C_PRD_DATA_CR_PRD_D_PH_3_MASK,
+              phase << I2C_PRD_DATA_CR_PRD_D_PH_3_SHIFT);
+}
+
+/****************************************************************************
+ * Name: bl602_set_i2c_clk
+ *
+ * Description:
+ *   set I2C clock.
+ *
+ ****************************************************************************/
+
+void bl602_set_i2c_clk(uint8_t enable, uint8_t div)
+{
+  modifyreg32(BL602_CLK_CFG3,
+              CLK_CFG3_I2C_CLK_DIV_MASK,
+              div << CLK_CFG3_I2C_CLK_DIV_SHIFT);
+
+  if (enable)
+    {
+      modifyreg32(BL602_CLK_CFG3, 0, CLK_CFG3_I2C_CLK_EN);
+    }
+  else
+    {
+      modifyreg32(BL602_CLK_CFG3, CLK_CFG3_I2C_CLK_EN, 0);
+    }
+}
+
+/****************************************************************************
+ * Name: bl602_i2c_clockset
+ *
+ * Description:
+ *   set i2c clock
+ *
+ ****************************************************************************/
+
+static void bl602_i2c_clockset(uint32_t clk)
+{
+  uint8_t bclk_div = 0;
+
+  bclk_div = bl602_glb_get_bclk_div();
+  if (clk >= 100000)
+    {
+      bl602_set_i2c_clk(1, 0);
+      bl602_i2c_setprd((SYS_TEMCORE_CLOCK_GET / (bclk_div + 1)) / (clk * 4) -
+                       1);
+    }
+  else if (clk >= 8000)
```

Comment:
Why are we supporting these lower frequencies?  I am not aware of an i2c device that operates so slow.  Usually you just have 100k 400k and some special devices that can go up to 5MHz

Reply:
BL602 I2c clk Ranges: 305Hz~3MHz

100~400K, test pass( use device)

3MHz, test pass (use oscilloscope )